rfc-087-use-libraries-where-possible.md

```diff
@@ -0,0 +1,98 @@
+# Use Existing Libraries Where Possible
+
+## Summary
+
+We should use existing open source libraries wherever possible,
+rather than writing our own code, but we should ensure that those
+libraries are trustworthy as best we can.
+
+## Problem
+
+Writing and maintaining code takes time, and any code is a possible
+source of bugs. When we implement our own solutions for problems
+that exist elsewhere, we may repeat work, increase development time,
+and increase the potential for error.
+
+## Proposal
+
+When we need to add new generic functionality to our code, we have a
+choice of whether to write our own implementation, or use an
+off-the-shelf open source library.
+
+By default, we SHOULD aim to find and reuse existing open source
+libraries that fulfil the requirement rather than writing our own.
+
+Using existing code is preferable in order to get better tested
+code, a lower maintenance burden, and more features than we might
+normally implement ourselves.
+
+Some examples might include:
+
+- Date handling and formatting
+- Language inflection
+- Queueing and background job running
+- Connecting to remote APIs
+
+The need for this will differ across languages and frameworks. For
+instance, Rails provides many capabilities that separate libraries
+would be needed for when working in Node.js. Conversely, frontend
+code may wish to avoid dependencies to keep file sizes low.
+
+Acquiring dependencies is not free, however. Developers MUST check
+them for license compatibility, and SHOULD assess them for quality
+and reliability, so that we are sure they will be a net positive
+addition to our projects.
+
+### License Compatibility
+
+Our code is MIT licensed by default. All dependencies MUST have
+compatible licensing. In particular this means looking out for
+accidental aquisition of a dependency on GPL software (or other
+[copyleft](https://en.wikipedia.org/wiki/Copyleft) licenses that put
+license conditions on derivatives).
+
+You can use tools to check license compatibility in your entire
+dependency tree:
+
+- Ruby: `bundle licenses | sed 's/.*: //g' | sort | uniq`
```

Comment:
Should this work as-is on any repo? Only it chucks an error on playbook & buy-for-your-school. Do we need more instructions?

Reply:
I wonder if it requires a certain version of bundler. I'll check, thanks!

Reply:
Hm it was added in 1.3.0, so it should be fine - I'll try it on playbook myself.

Reply:
```
playbook on  master [$?⇣] via ⬢ v14.6.0 via 💎 v2.7.2
❯ bundler -v
Bundler version 2.1.4

playbook on  master [$?⇣] via ⬢ v14.6.0 via 💎 v2.7.2
❯ bundle licenses | sed 's/.*: //g' | sort | uniq
standard: Unknown
Apache-2.0
BSD-2-Clause
BSD-3-Clause
CC0
MIT
Ruby
```

Works here, what happens for you?

Reply:
Weird, on playbook main I get
```
oreo:playbook laura$ bundle licenses | sed 's/.*: //g' | sort | uniq
Could not find kramdown-2.3.1 in any of the sources
```
```
oreo:playbook laura$ bundler -v
Bundler version 2.1.4
```

Reply:
Ugh, ignore me, it does work 🤦